README.md
# pokemon-cursor

An unofficial opensource Pokemon cursor theme for Windows and Linux.

[![build](https://github.com/ful1e5/pokemon-cursor/actions/workflows/build.yml/badge.svg)](https://github.com/ful1e5/pokemon-cursor/actions/workflows/build.yml)

## Pokemon Cursor needs your Input

Until 2021 my cursors projects were well funded by [pling.com](https://www.pling.com) but since the
[pling-factor](https://www.pling.com/terms/payout) on the website has decreased and monthly payments
are <500$, It is now dependent on community funding and sponsorships. If you want to help me to maintain
Pokemon Cursor and my other open source projects actively, consider sponsoring my work on [GitHub Sponsor](https://github.com/sponsors/ful1e5)
or DM me on [Twitter](https://twitter.com/ful1e5) if your company would like to support my projects,
I will gladly look into it and post your avatar in the project's README.

I appreciate all the wonderful people who patronize and sponsoring my work.

## Sponsors

<!-- Add your name or avatar here with the Pull Request in case I missed it. -->

N/A

---

![Pokemon](https://imgur.com/Xhm2xgS.png)

> **Note**
> All cursor's `.svg` files are found in [svg](./svg) directory or you can also find them on
> [Figma](https://www.figma.com/file/OoAFKT3u1HkzSe74QrLarQ/pokemon-cursor?node-id=0%3A1).

#### Cursor Sizes

<kbd>22</kbd>
<kbd>24</kbd>
<kbd>28</kbd>
<kbd>32</kbd>
<kbd>40</kbd>
<kbd>48</kbd>
<kbd>56</kbd>
<kbd>64</kbd>
<kbd>72</kbd>
<kbd>80</kbd>
<kbd>88</kbd>
<kbd>96</kbd>

## Cursor Sizes

### Xcursor Sizes:

<kbd>22</kbd>
<kbd>24</kbd>
<kbd>28</kbd>
<kbd>32</kbd>
<kbd>40</kbd>
<kbd>48</kbd>
<kbd>56</kbd>
<kbd>64</kbd>
<kbd>72</kbd>
<kbd>80</kbd>
<kbd>88</kbd>
<kbd>96</kbd>

### Windows Cursor Size:

- <kbd>16x16</kbd> - Small
- <kbd>24x24</kbd> - Regular
- <kbd>32x32</kbd> - Large
- <kbd>48x48</kbd> - Extra Large

#### Colors

![#FFCB05](https://imgur.com/Q3Y7xFo.png)
![#3C5AA6](https://imgur.com/rlKmamu.png)
![#0070B6](https://imgur.com/epcnFzB.png)
![#000000](https://imgur.com/ScBWhdE.png)
![#ffffff](https://imgur.com/AiR8P8t.png)

## How to get it

### Easiest Way

You can download latest `stable` & `development` releases from
[Release Page](https://github.com/ful1e5/pokemon-cursor/releases).

## Installing Pokemon Cursor

#### Linux/X11

**Installation:**

```bash
tar -xvf Pokemon.tar.gz                   # extract `Pokemon.tar.gz`
mv Pokemon ~/.icons/                      # Install to local users
sudo mv Pokemon /usr/share/icons/         # Install to all users
```

**Uninstallation:**

```bash
rm ~/.icons/Pokemon                       # Remove from local users
sudo rm /usr/share/icons/Pokemon          # Remove from all users
```

#### Windows

**Installation:**

1. Unzip `.zip` file
2. Open unziped directory in Explorer, and **right click** on `install.inf`.
3. Click 'Install' from the context menu, and authorize the modifications to your system.
4. Open Control Panel > Personalization and Appearance > Change mouse pointers,
   and select **Pokemon Cursors**.
5. Click '**Apply**'.

**Uninstallation:**

Run the `uninstall.bat` script packed with the `.zip` archive

**OR** follow these steps:

1. Go to **Registry Editor** by typing the same in the _start search box_.
2. Expand `HKEY_CURRENT_USER` folder and expand `Control Panel` folder.
3. Go to `Cursors` folder and click on `Schemes` folder - all the available custom cursors that are
   installed will be listed here.
4. **Right Click** on the name of cursor file you want to uninstall; for eg.: _Pokemon Cursors_ and
   click `Delete`.
5. Click '**yes**' when prompted.

## Build From Source

#### Notes

- Pokemon build configuration and cursor hotspot settings are bundled in the `build.toml` file.
- Check out the scripts section in [package.json](./package.json) to see how we build the cursor theme,
  excluding the render scripts. They are useful for converting `.svg` files to `.png` files.
- yarn is optional, For building XCursors and Windows cursors from `.png` files or resizing them
  you don't need that. If you want to develop/modify Pokemon's colors, and bitmaps, or generate a png
  file from a svg, Then you can use yarn because bitmapper is written in TypeScript.

### Build prerequisites

- Python version 3.7 or higher
- [clickgen](https://github.com/ful1e5/clickgen)>=2.1.2 (`pip install clickgen`)
- [yarn](https://github.com/yarnpkg/yarn)

### Quick start

1. Install [build prerequisites](#build-prerequisites) on your system
2. `https://github.com/ful1e5/pokemon-cursor`
3. `cd pokemon-cursor && yarn build`
4. See [Installing Pokemon Cursor](#installing-pokemon-cursor).

### Building

> **Note**
> Bitmaps are already generated in the `bitmaps` directory and **managed by the maintainer**
> (do not edit them directly).

First make sure you installed the [build prerequisites](#build-prerequisites).
Now that you have the dependencies, you can try build individual themes from bitmaps and
customize sizes, target platform, and etc. with the `ctgen` CLI (packed with `clickgen`).

#### `yarn build` aberration

Here are the default commands we used to build the Pokemon-cursor and packed them into `yarn build`:

```bash
ctgen build.toml
```

Afterwards, the themes can be found in the `themes` directory.

#### Customize Sizes

> **Note**
> You can change the cursor size up to 200 because pngs are rendered with 200x200.
> If the cursor is resized by more than rendered png size, the final cursor will be blurred.

##### Customize Windows Cursor size

To build Windows cursor with size `16`:

> **Warning**
> Windows cursor supports only one size, if multiple sizes are given with `-s` the first size will
> be considered in build.

```bash
ctgen build.toml -s 16 -p windows -c 'An unofficial Pokemon cursors with size 16'
```

You can also customize output directory with `-o` option:

```bash
ctgen build.toml -s 16 -p windows -o 'out' -c 'An unofficial Pokemon cursors with size 16'
```

##### Customize XCursor size

To build XCursor with size `16`:

```bash
ctgen build.toml -s 16 -p x11 -c 'An unofficial Pokemon cursors with size 16'
```

You can also assign multiple sizes to `ctgen` for XCursors build:

```bash
ctgen build.toml -s 16 24 32 -p x11 -c 'An unofficial Pokemon cursors with size 16'
```

# Bugs

Bugs should be reported [here](https://github.com/ful1e5/pokemon-cursor/issues) on the Github issues page.

# Getting Help

You can create a **issue**, I will help you.

# Contributing

Check [CONTRIBUTING.md](CONTRIBUTING.md), any suggestions for features and contributions to the continuing code masterelopment can be made via the issue tracker or code contributions via a `Fork` & `Pull requests`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

## [v2.0.0] - 23 October 2022

### Added

- tidy docs inside README.md
- Add cursor top_left_arrow cursor ful1e5/BreezeX_Cursor#10 ful1e5/BreezeX_Cursor#11
- Uninstall docs ful1e5/apple_cursor#79 ful1e5/apple_cursor#80
- feat: CLI for bitmapper #104
- doc: build docs updated

### Changed

- refactor: bitmapper moved to individual project [cbmp]()
- Removed copyright content from theme and tag this project as unofficial.
- prettier bitmapping logs

## [v1.0.0] - 08 February 2022

### Added

- Initial release 🎊
- Logo and badges
- CI/CD Pipelines

[unreleased]: https://github.com/ful1e5/pokemon-cursor/compare/v1.0.0...main
[v1.0.0]: https://github.com/ful1e5/pokemon-cursor/tree/v1.0.0


release.sh
#!/bin/bash
# A script for preparing binaries for version release of Pokemon Cursors, by Abdulkaiz Khatri

key="Pokemon"
comment="An unofficial Pokemon cursors."

# Cleanup old builds
rm -rf themes bin &&
ctgen build.toml -p x11 &&

# Building Pokemon Windows binaries
ctgen build.toml -p windows -s 16 -d "bitmaps/${key}" -n "${key}-Small" -c "$comment" &&
ctgen build.toml -p windows -s 24 -d "bitmaps/${key}" -n "${key}-Regular" -c "$comment" &&
ctgen build.toml -p windows -s 32 -d "bitmaps/${key}" -n "${key}-Large" -c "$comment" &&
ctgen build.toml -p windows -s 48 -d "bitmaps/${key}" -n "${key}-Extra-Large" -c "$comment" &&


# Compressing Binaries
mkdir -p bin &&
cd themes &&

tar -czvf "../bin/${key}.tar.gz" "${key}" &&
zip -rv "../bin/${key}-Windows.zip" "${key}-Small-Windows" "${key}-Regular-Windows" "${key}-Large-Windows" "${key}-Extra-Large-Windows" &&

cd ..


bitmapper/src/index.ts
import path from "path";

import { BitmapsGenerator, SVGHandler } from "./core";

const root = path.resolve(__dirname, "../../");
const svgDir = path.resolve(root, "svg");
const themeName = "Pokemon";

const main = async () => {
  console.log("Generating bitmaps for", themeName);

  const bitmapsDir = path.resolve(root, "bitmaps", themeName);
  const svg = new SVGHandler.SvgDirectoryParser(svgDir);

  const png = new BitmapsGenerator(bitmapsDir);
  const browser = await png.getBrowser();

  for (let { key, content } of svg.getStatic()) {
    console.log(" ==> Saving", key, "...");
    await png.generateStatic(browser, content, key);
  }

  for (let { key, content } of svg.getAnimated()) {
    console.log(" ==> Saving", key, "...");
    await png.generateAnimated(browser, content, key, { playbackRate: 1.0 });
  }

  await browser.close();
};

main();


builder/src/configure.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from typing import Any, Dict, Tuple, TypeVar

from clickgen.util import PNGProvider

from .constants import WIN_CURSORS_CFG, WIN_DELAY, X_CURSORS_CFG, X_DELAY

X = TypeVar("X")


def to_tuple(x: X) -> Tuple[X, X]:
    return (x, x)


def get_config(bitmaps_dir, **kwargs) -> Dict[str, Any]:
    """Return configuration of `Pokemon` pointers.

    :param bitmaps_dir: Path to .png file's directory.
    :type bitmaps_dir: ``str`` or  ``pathlib.Path``

    :param **kwargs:
        See below

    :Keyword Arguments:
        * *x_sizes* (``List[int]``) --
          List of pixel-sizes for xcursors.
        * *win_canvas_size* (``int``) --
          Windows cursor's canvas pixel-size.
        * *win_size* (``int``) --
          Pixel-size for Windows cursor.

    Example:

    ```python
        get_config(
            bitmaps_dir="./bitmaps",
            x_sizes=[24, 28, 32],
            win_canvas_size=32,
            win_size=24,
        )
    ```
    """

    w_size = to_tuple(kwargs.pop("win_size"))
    w_canvas_size = to_tuple(kwargs.pop("win_canvas_size"))
    raw_x_sizes = kwargs.pop("x_sizes")
    x_sizes = [to_tuple(size) for size in raw_x_sizes]
    png_provider = PNGProvider(bitmaps_dir)
    config: Dict[str, Any] = {}

    for key, item in X_CURSORS_CFG.items():
        x_hot: int = int(item.get("xhot", 0))
        y_hot: int = int(item.get("yhot", 0))
        hotspot: Tuple[int, int] = (x_hot, y_hot)

        delay: int = int(item.get("delay", X_DELAY))
        png = png_provider.get(key)
        if not png:
            raise FileNotFoundError(f"{key} not found")

        data = {
            "png": png,
            "x_sizes": x_sizes,
            "hotspot": hotspot,
            "delay": delay,
        }

        win_data = WIN_CURSORS_CFG.get(key)

        if win_data:
            win_key: str = str(win_data.get("to"))

            position: str = str(win_data.get("position", "center"))
            win_delay: int = int(win_data.get("delay", WIN_DELAY))

            canvas_size = win_data.get("canvas_size", w_canvas_size)
            win_size = win_data.get("size", w_size)

            # Because provided cursor size is bigger than cursor's canvas.
            # Also, "position" settings will not effect on cursor because the
            # cursor's canvas and cursor sizes are equals.
            if (win_size[0] > canvas_size[0]) | (win_size[1] > canvas_size[1]):
                canvas_size = win_size

            config[key] = {
                **data,
                "win_key": win_key,
                "position": position,
                "canvas_size": canvas_size,
                "win_size": win_size,
                "win_delay": win_delay,
            }
        else:
            config[key] = data

    return config


builder/src/constants.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from typing import Dict

# Info
AUTHOR = "Kaiz Khatri"
URL = "https://github.com/ful1e5/pokemon-cursor"

# XCursor
X_DELAY: int = 10

# Windows Cursor
WIN_DELAY = 1

X_CURSORS_CFG: Dict[str, Dict[str, int]] = {
    ##########
    # Static #
    ##########
    "all-scroll.png": {"xhot": 98, "yhot": 101},
    "bottom_left_corner.png": {"xhot": 96, "yhot": 97},
    "bottom_right_corner.png": {"xhot": 104, "yhot": 97},
    "bottom_tee.png": {"xhot": 95, "yhot": 123},
    "center_ptr.png": {"xhot": 92, "yhot": 42},
    "context-menu.png": {"xhot": 48, "yhot": 34},
    "copy.png": {"xhot": 48, "yhot": 34},
    "cross.png": {"xhot": 98, "yhot": 100},
    "crossed_circle.png": {"xhot": 48, "yhot": 34},
    "crosshair.png": {"xhot": 100, "yhot": 100},
    "dnd_no_drop.png": {"xhot": 100, "yhot": 100},
    "dotbox.png": {"xhot": 100, "yhot": 100},
    "hand1.png": {"xhot": 101, "yhot": 50},
    "hand2.png": {"xhot": 75, "yhot": 40},
    "left_ptr.png": {"xhot": 48, "yhot": 34},
    "left_side.png": {"xhot": 102, "yhot": 97},
    "left_tee.png": {"xhot": 76, "yhot": 99},
    "link.png": {"xhot": 48, "yhot": 34},
    "ll_angle.png": {"xhot": 66, "yhot": 142},
    "lr_angle.png": {"xhot": 126, "yhot": 137},
    "move.png": {"xhot": 98, "yhot": 53},
    "pencil.png": {"xhot": 63, "yhot": 151},
    "plus.png": {"xhot": 97, "yhot": 99},
    "question_arrow.png": {"xhot": 48, "yhot": 34},
    "right_ptr.png": {"xhot": 141, "yhot": 35},
    "right_tee.png": {"xhot": 119, "yhot": 98},
    "sb_down_arrow.png": {"xhot": 99, "yhot": 123},
    "sb_h_double_arrow.png": {"xhot": 98, "yhot": 101},
    "sb_left_arrow.png": {"xhot": 72, "yhot": 97},
    "sb_right_arrow.png": {"xhot": 139, "yhot": 101},
    "sb_up_arrow.png": {"xhot": 99, "yhot": 65},
    "sb_v_double_arrow.png": {"xhot": 98, "yhot": 99},
    "top_side.png": {"xhot": 101, "yhot": 95},
    "top_tee.png": {"xhot": 101, "yhot": 96},
    "ul_angle.png": {"xhot": 78, "yhot": 77},
    "ur_angle.png": {"xhot": 124, "yhot": 78},
    "vertical-text.png": {"xhot": 100, "yhot": 97},
    "wayland-cursor.png": {"xhot": 100, "yhot": 100},
    "X_cursor.png": {"xhot": 100, "yhot": 100},
    "xterm.png": {"xhot": 101, "yhot": 93},
    "zoom-in.png": {"xhot": 82, "yhot": 82},
    "zoom-out.png": {"xhot": 82, "yhot": 82},
    ############
    # Animated #
    ############
    # Note: Animated cursors don't need an extension and frame numbers.
    "left_ptr_watch": {"xhot": 48, "yhot": 34},
    "wait": {"xhot": 100, "yhot": 100},
}

WIN_CURSORS_CFG: Dict[str, Dict[str, str]] = {
    ##########
    # Static #
    ##########
    "right_ptr.png": {"to": "Alternate", "position": "top_left"},
    "cross.png": {"to": "Cross"},
    "left_ptr.png": {"to": "Default", "position": "top_left"},
    "bottom_right_corner.png": {"to": "Diagonal_1"},
    "bottom_left_corner.png": {"to": "Diagonal_2"},
    "pencil.png": {"to": "Handwriting"},
    "question_arrow.png": {"to": "Help", "position.png": "top_left"},
    "sb_h_double_arrow.png": {"to": "Horizontal"},
    "xterm.png": {"to": "IBeam", "position": "top_left"},
    "hand2.png": {"to": "Link", "position": "top_left"},
    "hand1.png": {"to": "Move"},
    "crossed_circle.png": {"to": "Unavailiable", "position": "top_left"},
    "sb_v_double_arrow.png": {"to": "Vertical"},
    ############
    # Animated #
    ############
    # Note: Animated cursors don't need frame numbers.
    "left_ptr_watch": {"to": "Work", "position": "top_left"},
    "wait": {"to": "Busy"},
}


builder/src/generator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from pathlib import Path
from typing import Any, Dict, NamedTuple

from clickgen.builders import WindowsCursor, XCursor
from clickgen.core import CursorAlias
from clickgen.packagers import WindowsPackager, XPackager

from .constants import AUTHOR, URL
from .symlinks import add_missing_xcursor


class Info(NamedTuple):
    """Theme basic information.

    :param name: Theme title.
    :type name: ``str``

    :param comment: quick information about theme.
    :type comment: ``str``
    """

    name: str
    comment: str


def xbuild(config: Dict[str, Dict[str, Any]], x_out_dir: Path, info: Info) -> None:
    """Build `Pokemon` cursor theme for only `X11`(UNIX) platform.

    :param config: `Pokemon` configuration.
    :type config: ``Dict``

    :param x_out_dir: Path to the output directory,\
                Where the `X11` cursor theme package will generate.\
                It also creates a directory if not exists.
    :type x_out_dir: ``pathlib.Path``

    :param info: Content theme name & comment.
    :type info: Info
    """

    for _, item in config.items():
        with CursorAlias.from_bitmap(item["png"], item["hotspot"]) as alias:
            x_cfg = alias.create(item["x_sizes"], item["delay"])

            print(f"Building '{x_cfg.stem}' XCursor...")
            XCursor.create(x_cfg, x_out_dir)

    add_missing_xcursor(x_out_dir / "cursors")
    XPackager(x_out_dir, info.name, info.comment)


def wbuild(config: Dict[str, Dict[str, Any]], win_out_dir: Path, info: Info) -> None:
    """Build `Pokemon` cursor theme for only `Windows` platforms.

    :param config: `Pokemon` configuration.
    :type config: ``Dict``

    :param win_out_dir: Path to the output directory,\
                  Where the `Windows` cursor theme package will generate.\
                  It also creates a directory if not exists.
    :type win_out_dir: ``pathlib.Path``

    :param info: Content theme name & comment.
    :type info: Info
    """

    for _, item in config.items():
        with CursorAlias.from_bitmap(item["png"], item["hotspot"]) as alias:
            alias.create(item["x_sizes"], item["delay"])

            if item.get("win_key"):
                win_cfg = alias.reproduce(
                    size=item["win_size"],
                    canvas_size=item["canvas_size"],
                    position=item["position"],
                    delay=item["win_delay"],
                ).rename(item["win_key"])

                print(f"Building '{win_cfg.stem}' Windows Cursor...")
                WindowsCursor.create(win_cfg, win_out_dir)

    WindowsPackager(win_out_dir, info.name, info.comment, AUTHOR, URL)


def build(
    config: Dict[str, Dict[str, Any]], x_out_dir: Path, win_out_dir: Path, info: Info
) -> None:
    """Build `Pokemon` cursor theme for `X11` & `Windows` platforms.

    :param config: `Pokemon` configuration.
    :type config: ``Dict``

    :param x_out_dir: Path to the output directory,\
                Where the `X11` cursor theme package will generate.\
                It also creates a directory if not exists.
    :type x_out_dir: ``pathlib.Path``

    :param win_out_dir: Path to the output directory,\
                  Where the `Windows` cursor theme package will generate.\
                  It also creates a directory if not exists.
    :type win_out_dir: ``pathlib.Path``

    :param info: Content theme name & comment.
    :type info: Info
    """

    for _, item in config.items():

        with CursorAlias.from_bitmap(item["png"], item["hotspot"]) as alias:
            x_cfg = alias.create(item["x_sizes"], item["delay"])

            print(f"Building '{x_cfg.stem}' XCursor...")
            XCursor.create(x_cfg, x_out_dir)

            if item.get("win_key"):
                win_cfg = alias.reproduce(
                    size=item["win_size"],
                    canvas_size=item["canvas_size"],
                    position=item["position"],
                    delay=item["win_delay"],
                ).rename(item["win_key"])

                print(f"Building '{win_cfg.stem}' Windows Cursor...")
                WindowsCursor.create(win_cfg, win_out_dir)

    add_missing_xcursor(x_out_dir / "cursors")
    XPackager(x_out_dir, info.name, info.comment)

    WindowsPackager(win_out_dir, info.name, info.comment, AUTHOR, URL)


builder/build.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
from pathlib import Path

from src.configure import get_config
from src.generator import Info, build, wbuild, xbuild

parser = argparse.ArgumentParser(
    prog="pokemon_cursor_builder",
    description="'Pokemon' cursor build python script.",
)


# Positional Args.
parser.add_argument(
    "platform",
    choices=("windows", "unix", "all"),
    default="all",
    const="all",
    nargs="?",
    help="Set package type, Which you want to build. (default: '%(default)s')",
)


# Optional Args.
parser.add_argument(
    "-p",
    "--png-dir",
    dest="png_dir",
    metavar="PNG",
    type=str,
    default="../bitmaps",
    help="To change pngs directory. (default: %(default)s)",
)

parser.add_argument(
    "-o",
    "--out-dir",
    dest="out_dir",
    metavar="OUT",
    type=str,
    default="../themes",
    help="To change output directory. (default: %(default)s)",
)


parser.add_argument(
    "-xs",
    "--xsizes",
    dest="xsizes",
    metavar="INT",
    nargs="+",
    default=[
        22,
        24,
        28,
        32,
        40,
        48,
        56,
        64,
        72,
        80,
        88,
        96,
    ],
    type=int,
    help="Set pixel-size for xcursor. (default: %(default)s)",
)


parser.add_argument(
    "-ws",
    "--win-size",
    dest="win_size",
    metavar="INT",
    default=24,
    type=int,
    help="Set pixel-size for Windows cursors. (default: %(default)s)",
)


parser.add_argument(
    "-wcs",
    "--win-canvas-size",
    dest="win_canvas_size",
    metavar="INT",
    default=32,
    type=int,
    help="Set pixel-size for Windows cursor's canvas. (default: %(default)s)",
)

# Preparing build
args = parser.parse_args()

bitmaps_dir = Path(args.png_dir)
name = bitmaps_dir.stem

x_out_dir = Path(args.out_dir) / name
win_out_dir = Path(args.out_dir) / f"{name}-Windows"

# Windows Canvas & Cursor sizes
win_size: int = args.win_size
win_canvas_size: int = args.win_canvas_size
if win_canvas_size < win_size:
    win_canvas_size = win_size

print(f"Getting '{name}' bitmaps ready for build...")

config = get_config(
    bitmaps_dir,
    x_sizes=args.xsizes,
    win_canvas_size=args.win_canvas_size,
    win_size=args.win_size,
)

info = Info(name=name, comment=f"{name} Cursors")

if args.platform == "unix":
    xbuild(config, x_out_dir, info)
elif args.platform == "windows":
    wbuild(config, win_out_dir, info)
else:
    build(config, x_out_dir, win_out_dir, info)
